processing/src/main/java/io/druid/segment/DimensionSelectors.java

```diff
@@ -0,0 +1,458 @@
+        final ArrayList<Integer> vals = Lists.newArrayList();
+        if (dimensionValues != null) {
+          for (int i = 0; i < dimensionValues.size(); ++i) {
+            vals.add(i);
+          }
+        }
+
+        return new IndexedInts()
+        {
+          @Override
+          public int size()
+          {
+            return vals.size();
+          }
+
+          @Override
+          public int get(int index)
+          {
+            return vals.get(index);
+          }
+
+          @Override
+          public Iterator<Integer> iterator()
+          {
+            return vals.iterator();
+          }
+
+          @Override
+          public void close() throws IOException
+          {
+
+          }
+
+          @Override
+          public void fill(int index, int[] toFill)
+          {
+            throw new UnsupportedOperationException("fill not supported");
+          }
+        };
+      }
+
+      @Override
+      public int getValueCardinality()
+      {
+        return DimensionSelector.CARDINALITY_UNKNOWN;
+      }
+
+      @Override
+      public String lookupName(int id)
+      {
+        final String value = rowSupplier.get().getDimension(dimension).get(id);
+        return extractionFn == null ? value : extractionFn.apply(value);
+      }
+
+      @Override
+      public int lookupId(String name)
+      {
+        if (extractionFn != null) {
+          throw new UnsupportedOperationException("cannot perform lookup when applying an extraction function");
+        }
+        return rowSupplier.get().getDimension(dimension).indexOf(name);
+      }
+    };
+  }
+
+  public static DimensionSelector makeMultiDimensionalSelector(
+      final List<DimensionSelector> selectors,
+      final ExtractionFn extractionFn
+  )
+  {
+    Preconditions.checkArgument(extractionFn != null && (extractionFn.arity() < 0 || extractionFn.arity() == selectors.size()),
+        "ExtractionFn should be specified and have the same arity as the number of dimensions");
+
+    return new DimensionSelector()
+    {
+      StringDimensionIndexer.DimensionDictionary dimLookup =
+          new StringDimensionIndexer.DimensionDictionary();
+
+      @Override
+      public IndexedInts getRow()
+      {
+        final List<IndexedInts> rows = Lists.transform(
+            selectors,
+            new Function<DimensionSelector, IndexedInts>()
+            {
+              @Override
+              public IndexedInts apply(DimensionSelector selector)
+              {
+                return selector.getRow();
+              }
+            }
+        );
+
+        int size = 1;
+        for (IndexedInts row: rows) {
+          size *= row.size();
+        }
+        final int finalSize = size;
+
+        final List<Integer> vals = Lists.newArrayListWithCapacity(size);
+
+        for (int idx = 0; idx < size; idx++) {
+          List<String> args = Lists.newArrayListWithCapacity(rows.size());
```

Comment:
(Comment unrelated to the above)

DimensionSelector is single-threaded (I suppose?), so this `args` list could be cached in a field of this DimensionSelector, and just `cleared()` on each getRow() instead of new allocation.